sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -422,47 +422,487 @@ case class Cast(child: Expression, dataType: DataType) extends UnaryExpression w
   protected override def nullSafeEval(input: Any): Any = cast(input)
 
   override def genCode(ctx: CodeGenContext, ev: GeneratedExpressionCode): String = {
-    // TODO: Add support for more data types.
-    (child.dataType, dataType) match {
+    val eval = child.gen(ctx)
+    val nullSafeCast = nullSafeCastFunction(child.dataType, dataType, ctx)
+    if (nullSafeCast != null) {
+      eval.code +
+        castCode(ctx, eval.primitive, eval.isNull, ev.primitive, ev.isNull, dataType, nullSafeCast)
+    } else {
+      super.genCode(ctx, ev)
```

Comment:
castCode is intendedly take all these parameters because we have to do cast for `ComplexType` as well, so we need call castCode recursively in these casts.